### PR TITLE
Add cool bar style percentage view

### PR DESCRIPTION
**Pull request by rashil2000**
image

I'll add 'New-PercentageBar' in the optional dependencies section of the wiki.

I have renamed the 'process' function to 'cpu_usage' to remove any ambiguity at all (and also to keep parity with Neofetch).

### CODE REVIEW

**Review by jcwillox — CHANGES_REQUESTED** [review]
This is a great addition but I'm not sure about the implementation, it's not necessary to add a new dependency just for progress bars, they are simple enough to implement within winfetch. We should also add more options to show the progress percent, and to swap which side the bar is on (like neofetch), this could be done by adding a new function to generate progress bar info segments, i.e you pass it the `$percent` and `$summary`, it then looks at some global variables to decide how to format it including formatting it without progress bar. This would reduce redundant code as you've duplicated the `$content =` parts.

Here are some example progress bar drawing functions I coded up, `draw_progress` was just a simple 2-liner I wrote, the `draw_progress2` function is more advanced and has the different colored blocks for higher percentages.

```pwsh
$e = [char]27
$length = 10

function draw_progress($percent) {
    $blocks = [math]::round($percent / 100 * $length)
    return "[$e[32m " + "$([char]9632)" * $blocks + "$e[0m" + "-" * ($length - $blocks) + " ]"
}

function draw_progress2($percent) {
    $blocks = [math]::round($percent / 100 * $length)
    $empty = $length - $blocks

    $bar = "[ $e[32m"

    for ($i = 1; $i -le $blocks; $i++) {
        if ($i -eq 7) {
            $bar += "$e[93m"
        } elseif ($i -eq 9) {
            $bar += "$e[91m"
        }
        $bar += "$([char]9632)"
    }

    return $bar + "$e[0m" + "-" * $empty + " ]"
}

# print progress bars
for ($i = 0; $i -le 100; $i+=10) {
    draw_progress $i
}

for ($i = 0; $i -le 100; $i+=10) {
    draw_progress2 $i
}
```

**Comment by jcwillox** [comment]
Another thing is that we should still show the `value/maximum`, the percentage bars are only a replacement for the actual `43%` otherwise there is no way to tell the maximum value.

**Comment by rashil2000** [comment]
I was under the assumption that the script needs to be minimal and as small in size as possible (remember, an important USP for winfetch was that it isn't as bloated as neofetch). Showing percentage bars is not a core functionality, that's why I made it optional. The 'New-PercentageBar' script itself is highly configurable, FWIW.

> reduce redundant code as you've duplicated the $content = parts

As for this, I can make a simple splatting function to handle this, shouldn't be a concern.

**Comment by rashil2000** [comment]
> Another thing is that we should still show the `value/maximum`, the percentage bars are only a replacement for the actual `43%` otherwise there is no way to tell the maximum value.

I think this then defeats the purpose of a 'bar'. It's purpose is to show relative values; if `used` part is shown, the `total` part is implicit from the amount that the bar is filling up.

**Comment by jcwillox** [comment]
> I think this then defeats the purpose of a 'bar'. Its purpose is to show relative values; if `used` part is shown, the `total` part is implicit from the amount that the bar is filling up.

That's like saying the 'total' part is implicit when we show `(48%)` because I know that's half, a bar is an easy way to visualise a percentage sure, but you may still want that detail

> I was under the assumption that the script needs to be minimal and as small in size as possible (remember, an important USP for winfetch was that it isn't as bloated as neofetch). Showing percentage bars is not a core functionality, that's why I made it optional. The 'New-PercentageBar' script itself is highly configurable, FWIW.

Yes, however, its a relatively popular feature, that can be implemented in **17 lines** (or even 4)! like why go add an external dependency and complicate things for everyone, when adding that percentage bar function will have basically 0 effect on install or runtime speed.

The issue isn't really size or minimalism, we want a robust program that is simple to use and most importantly fast. That's the big issue with neofetch on windows, it's slow and doesn't work well.

**Comment by jcwillox** [comment]
Here's some draft code/ideas on a way we could add much more customizability to the percentage bar/segments and also reduce redundancy

```pwsh
$PercentageBar = $true
$ShowPercentage = $true
$ShowSummary = $true

# an additional option to control which segments use the bar
$ShowPercentageBars = @("cpu_usage", "memory", "disk")

# possible idea for ordering
$BarSegmentOrder = @("bar", "percent", "summary")

# a basic implementation of a function to create segments
function percentage_segment($percent, $summary) {
    $output = ""
    if ($PercentageBar) {
        $output += draw_percentage2 $percent
    }
    if ($ShowSummary) {
        $output += $summary
    }
    if ($ShowPercentage) {
        $output += "($percent%)"
    }
    return $output
}

return @{
    title   = "Memory"
    content = percentage_segment $usage ("{0:f1} GiB / {1:f1} GiB" -f $used,$total)
}
```

**Comment by rashil2000** [comment]
If you have the bar function ready, feel free to commit directly to this branch

**Comment by rashil2000** [comment]
@jcwillox I have added a builtin function like you said.

**Comment by rashil2000** [comment]
@jcwillox I've added the config options, following Neofetch

**Comment by jcwillox** [comment]
Great, I'll check it out. 

I was going to say we should probably rename the `$percentagebar` parameter to `-showpercentagebars` or `-showbars` and have it work similar to other parameters like `-showdisks`. Instead of accepting true/false it should accept and array of info segment names or a `*` for all possible segments. e.g. `winfetch -showpercentagebars *` or `winfetch -showbars cpu_usage,disk`. This way users can quickly control which info segments show the percentage bar.

Also those function names should be camel-case e.g. `getpercentbar -> get_percent_bar` for consistency.

**Review by jcwillox — CHANGES_REQUESTED** [review]
I'm assuming the codes not done because using the `-percentagebar` parameter doesn't do anything. Also, the default style for each info row should not change, e.g. `188 GiB / 231 GiB (81%)` changed to `81% (188 GiB / 231 GiB)`. I assume you did that for simplicity to align it with the `cpu_load` / `battery` segment, but it's not necessary. You can change `$text` so its just the info however that segment wants to display it, and `$percentage` is just used for `get_percent_bar`. I don't know how set you are on `printlevel` but it's a little vague, the name I originally thought of was `get_bar_segment`.

You've got two conflicting options here, `-percentagebar` and `$cpustyle` these won't work together, my idea was to have a variable to enable the display of the bar per segment (`-showbars`) and then another option to globally control the display order of these percentage bar segments (`$BarSegmentOrder`). Either option is viable but if we're going to add per segment styling then we need to add parameters for each segment like neofetch does e.g. `-cpustyle infobar`.

Below is a modified version of `printlevel` that takes into account the things I mentioned, this would probably be fine for now, but it could be made more robust, to handle, for example, the location of percent element. I also included a `barprefix` option if you want to align the percentage bars when using `barinfo`.

```pwsh
function get_bar_segment {
    param (
        [string]$barprefix,
        [string]$style,
        [int]$percentage,
        [string]$info
    )

    switch ($style) {
        "bar" { return "$(get_percent_bar $percentage)" }
        "infobar" { return "$info $(get_percent_bar $percentage)" }
        "barinfo" { return "$barprefix$(get_percent_bar $percentage) $info" }
        default { return $info }
    }
}

get_bar_segment "" $cpustyle $loadpercent "$loadpercent% ($proccount processes)"

get_bar_segment "" $diskstyle $usage "$(to_units $used) / $(to_units $total) ($usage%)"

get_bar_segment "  " $batterystyle $battery.EstimatedChargeRemaining "$($battery.EstimatedChargeRemaining)% ($status$timeFormatted)"
```

File: winfetch.ps1 [hunk 1]
```diff
@@ -178,6 +218,16 @@ $defaultConfig = @'
...
+# Configure how to show info for levels
+# 0 is for text only. (Default)
+# 1 is for bar only.
+# 2 is for text + bar.
+# 3 is for bar + text.
+# $cpustyle = 1
+# $memorystyle = 2
+# $diskstyle = 3
+# $batterystyle = 3
+
```
Comment (jcwillox) [inline winfetch.ps1:230]: We can use strings here and in the switch on line 135, that'd be much easier for users and wouldn't require as much explanation.
e.g. $cpustyle = "infobar"

**Comment by rashil2000** [comment]
I have made the required changes. Let me know if this is good to go.

**Comment by jcwillox** [comment]
Great, almost, can you make it so that the default style is the same as before, the code block in my last comment provides an example implementation.

**Comment by rashil2000** [comment]
There's one issue with your codeblock - it prints _both_ the percentage and the bar in some cases - which defeats the purpose of having a bar (or more precisely, makes the bar redundant).

**Comment by rashil2000** [comment]
I have changed the order of default text, this makes it same as it used to be before.

**Comment by jcwillox** [comment]
> There's one issue with your codeblock - it prints _both_ the percentage and the bar in some cases - which defeats the purpose of having a bar (or more precisely, makes the bar redundant).

I get why you say that, but it's not really true, a progress bar is just another way display data, all three methods show the same information but represent it differently, each method comes with its own pros/cons and is better for certain use cases. And in general, most progress bars are accompanied by their percentage full.

Currently, we have two representations that we use, one for cpu/battery `$percent% ($info)` and one for memory/disk `$info ($percent%)`. Assuming we don't want to go the route of what I suggested then we could add a parameter to the `get_level_info` function to switch to using the alternate style. Possibly in the future we could even make the `percent` a field users can control e.g. `$cpudisplay = bar,text,percent`

**Comment by rashil2000** [comment]
>we could add a parameter to the `get_level_info` function to switch to using the alternate style

I was thinking the exact same thing, but I didn't feel the need to complicate things (as of now) to a point where _5 parameters_ are needed to specify a function that's needed just by 4 segments. 

>we could even make the percent a field users can control e.g. `$cpudisplay = bar,text,percent`

That's a nice suggestion, we can do it in the future, however it's not a blocker per say for this PR so I'm going to merge this for now.

**Pull request merged**